### DISCUSSION TITLE
Embedding Node.js ESM

### DISCUSSION THREAD

**Discussion by segevfiner** [nodejs/help#4607] [Q&A]
The current embedding API gives you to ability to run a CJS script (That seems to need to bootstrap `require` by itself), or to get a callback with a `runCjs` function, but what's the right way to run ESM when embedding? The current sample https://nodejs.org/docs/latest-v20.x/api/embedding.html uses `runInThisContext` which doesn't support dynamic import by default without `vm.constants.USE_MAIN_CONTEXT_DEFAULT_LOADER` which dumps an experimental warning, and `vm.SourceTextModule` is experimental, which I'm not sure how to correctly enable when embedding, and also seems incomplete such as requiring a `link` callback without giving you access to the default resolution algorithm, and so on... And the CJS main execution is sync, so if you `import`, how would you wait for the resulting promise to finish from embedding code?

**Comment by DavidPeicho** [comment]
I worked around that by doing something like that in the warmup script:

```js
globalThis._loadModuleESM = async function (source, url) {
    const {loadESM} = require('internal/process/esm_loader');

    let result = null;
    return loadESM((loader) => {
        result = loader.eval(source, url).catch((e) => {
            /* Wrap non-native errors to simplify error-detection. */
            if (!(e instanceof Error)) e = new Error(e);
            return e;
        });
        return result;
    }).then(() => result);
};
```

Then you can retrieve `_loadModuleESM` to load ESM modules on the C++ side. Don't forget to call this method using `MakeCallback`, and to spin the event loop.

It's not the best, but at least it works.
